Add feedback forms

wioux opened the pull request:
Adds a page at /feedback with a user satisfaction survey. Also adds a little drawer on the right side of the page for users to submit quick feedback on the current page.

Manage the /feedback survey questions at /admin/survey_questions?q%5Bsurvey_equals%5D=/feedback. Technically these should be reorderable but that isn't implemented yet...

vbrown608 commented on app/controllers/feedback_controller.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  def survey_questions
+    SurveyQuestion.
```
This is used in the helper too. A named scope for grabbing each appropriately-ordered survey might be nice.

wioux replied:
Good idea.

vbrown608 commented on app/admin/survey_question.rb:
```diff
@@ -0,0 +1,46 @@
+    f.inputs do
+      f.input :prompt
+
+      f.has_many :options, label: "xyz", sortable: :position, allow_destroy: true do |o|
```
What does `label: "xyz"` do?

wioux replied:
Whoops... it was supposed to remind me to circle back here and figure out how to change the label on a has_many fieldset.

vbrown608 commented on app/views/feedback/create.html.erb:
```diff
@@ -0,0 +1,2 @@
+<h1>Feedback#create</h1>
+<p>Find me in app/views/feedback/create.html.erb</p>
```
Should this file be removed?

wioux replied:
Yes, thanks.

vbrown608 commented on app/models/survey_response.rb:
```diff
@@ -0,0 +1,14 @@
+  private
+
+  def copy_prompt_from_question
+    self.prompt = survey_question.prompt
```
Could we delegate prompt to survey_question, or does this need to be saved to the response itself?

wioux replied:
The response saves both the question that was asked and the answer. This is in case the survey gets edited later and the question changes. It's a bit of a speculative concern :man_shrugging:. Instead I could have made the question prompts non-editable which would have made sense too. I'm not sure whether one approach is better.

vbrown608 replied:
Ahh got it. In that case I think this is better. Not being able to edit a small typo sounds frustrating.

vbrown608 commented on app/helpers/feedback_helper.rb:
```diff
@@ -0,0 +1,42 @@
+  def survey_prompt(question)
+    question.prompt.strip.tap do |prompt|
+      prompt << ":" unless prompt =~ /[.?!:]$/
+      prompt << "*" if question.required?
```
Could also be a css :after pseudoelement based on the required attribute.

wioux replied:
:+1:

wioux commented:
Great feedback, thanks! I'll make some changes and let you know when this is ready.

eff-soraya commented:
First off, I wanted to chime in with the unhelpful following comment:
You all are great! I really love how this is coming along. Thank you <3


Also, perhaps more helpful comments:

+ I wanted to second a bunch of points Vivian made...

    + I'm not sure if the feedback form needs to be on the front page,
    but I don't mind that it's there either. Maybe we could do an
    informal user test with someone on a different team and get their
    rough sense?

    + The bottom of the feedback drawer form appears to be cut off in my
    browser (Mozilla Firefox) — image attached for reference.

    + The feedback drawer seems to have a bit of a glitch on the Lessons
    page.

    + I wonder what a mobile version of the Feedback form might look
    like. Perhaps it could be at the bottom of a post?

    + I wonder if the "contact" page at the footer could also lead to
    the larger feedback page, rather than the present sec-feedback email.

    + I wonder if we could incorporate some kind of user flow for
    completing the more comprehensive survey.

    + I agree on the accessibility point and so appreciate that you
    caught that!


+ Separately...

    + I'd be keen to learn more about the details of how the feedback
    form works. As the recipient of the feedback, would I see which
    pages people commented on? What information would be coming in?



On 3/1/18 12:58 PM, V Brown wrote:
>
> *@vbrown608* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In app/models/survey_response.rb
> <https://github.com/EFForg/sec/pull/418#discussion_r171691484>:
>
> > @@ -0,0 +1,14 @@
> +class SurveyResponse < ApplicationRecord
> +  belongs_to :feedback
> +  belongs_to :survey_question
> +
> +  validates_presence_of :value, if: ->{ survey_question.required? }
> +
> +  before_create :copy_prompt_from_question
> +
> +  private
> +
> +  def copy_prompt_from_question
> +    self.prompt = survey_question.prompt
>
> Ahh got it. In that case I think this is better. Not being able to
> edit a small typo sounds frustrating.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/EFForg/sec/pull/418#discussion_r171691484>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AP73e0Y8DoUiS_JhfjKsz_vSQDkRnEGgks5taGDagaJpZM4SXYQi>.
>

-- 
Learn more about digital security at https://ssd.eff.org/.
Want to teach digital security? Check out the Security Education Companion at https://sec.eff.org/.
Check out tools for encrypting the web at https://www.eff.org/encrypt-the-web

wioux commented:
Notes from Soraya:

   * Save whether feedback was submitted from mobile or desktop.
   * Add a way to see how answers have changed over time (or limit to only "recent" responses).
   * Link to feedback survey from the bottom of lessons / articles.
   * Quick feedback drawer is cut off on her (MBPr) display.
   * How can we share this with the advisory group before release?

wioux commented:
I fixed the accessibility and style issues, nixed the homepage quick survey, and improved the admin form as suggested. I added the other comments Soraya had to #385 as I'd like to follow up on them in a separate branch, so this PR is ready for review again.
